### ISSUE TITLE
Scoped CSS style

### ISSUE THREAD

**Issue by msaraiva** [surface-ui/surface#1]
1. The first thing we need is to find a way to parse/change css. The only lib I could find was this one https://github.com/acammack-r7/erlang-css, but I'm afraid it hasn't been maintained for a while. So we either find an alternative or we need to roll our own parser. Maybe using `nimble_parsec` again?
2. Read the `scoped` attribute from `<style scoped>...</style>`, if it's set, add a unique component hash id to selectors and translate the code accordingly. Example:

    ```jsx
    <style scoped>
      .example {
        color: red;
      }
    </style>

    <div class="example">
      ...
    </div>
    ```
    becomes:
    ```jsx
    <style scoped>
      .example[data-sface-d4f9a23] {
        color: red;
      }
    </style>

    <div data-sface-d4f9a23>
      ...
    </div>
    ```
3. Keep track of styles definitions in order to render only one definition for each type of component. Currently, if we render more than one instance of a component, the style is also rendered, causing unnecessary duplication.

**Comment by Zurga** [comment]
Collecting the styles linked to the components is easily done using an Agent or by setting the styles as an attribute on the Component. The work (as far as I can tell) lies in injecting the relevant styles where they are needed. It would be necessary to collect a full tree of components and have the top-most component retrieve all the relevant styles of its children. 

Why do you see the need to parse the CSS, could this not be implemented in an agnostic manner? The de-duplication of styles would not need any parsing of the CSS, as far as I can tell.

At the moment, the the scoped parameter is [deprecated](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/style#Deprecated_attributes), although people are [trying to start it back up again](https://github.com/w3c/csswg-drafts/issues/3547). Would it not be better to advise people to keep their styles out of the components? Seems a bit cleaner to me, but I guess people have use cases for this.

Please let me know if you have any suggestions on how to tackle this, would love to help.

**Comment by msaraiva** [comment]
Hey @Zurga!

> Collecting the styles linked to the components is easily done using an Agent or by setting the styles as an attribute on the Component. The work (as far as I can tell) lies in injecting the relevant styles where they are needed. It would be necessary to collect a full tree of components and have the top-most component retrieve all the relevant styles of its children.

Yeah, that's what I have in mind so far.

> Why do you see the need to parse the CSS, could this not be implemented in an agnostic manner? The de-duplication of styles would not need any parsing of the CSS, as far as I can tell.

The parsing is needed in order to change the styles by adding a unique hash to each of the selectors as I tried to show above. That way, the user can create something like "private" css classes for a particular component without leaking those styles to any of the child components. You can see the same approach being used by `Vue.js` here https://vue-loader.vuejs.org/guide/scoped-css.html.

> At the moment, the the scoped parameter is deprecated, although people are trying to start it back up again. 

Since we have our own parser, we don't need to worry much about that. Surface is a higher-level abstraction on top of HTML. We can certainly have our own semantics for the `scoped` attribute or even use another attribute (or directive) to avoid confusion. Personally, I don't like much the idea of injecting css styles directly in the component's file. I would prefer to add a collocated `.css` file that would be automatically "scoped"and  injected in the generated code as described here https://github.com/msaraiva/surface/issues/22

> Would it not be better to advise people to keep their styles out of the components? Seems a bit cleaner to me, but I guess people have use cases for this.

Yeah, that's exactly why this issue was opened. I want to get as much feedback as possible about the pros and cons of having this feature in Surface. So far, I'm not convinced that it will make a big difference, that's why it's not even planned to be implemented for the first stable version. However, as far as I know, people use this a lot in `Vue` and `Svelte` and they love it, so let's wait and see if we get convinced that we should do this.

Thanks for your inputs!

**Comment by garretttaco** [comment]
I am very much so excited about this. 

Another take on styling could be following the approach of what [Styled Components](https://styled-components.com/) is for React. We could have this as a new library or as part of Surface. 

One of the main benefits of styled-components is being able to have dynamic CSS, which can take component props and update the CSS accordingly, and allows you to run Elixir code to calculate CSS. This also opens up the style system to allow helper libraries from the community: These may look similar to https://polished.js.org/, and https://github.com/styled-system/styled-system 

StyledSurface (or whatever you would want to call it 🙂) could learn from the internals of StyledComponents on when and how to append to the DOM with updated styles. We would then, if I am not mistaken could leverage LiveView to broadcast any style updates to a style tag at the bottom of the HTML head tag. The auto-generated classes would work the same as with the original scoped styles idea above.

What are your thoughts on this idea @msaraiva? I would be willing to work on a prototype to get some ideas flowing, with your guidance.

**Comment by msaraiva** [comment]
Hi @garretttaco!

It looks like there are so many different approaches to deal with CSS nowadays that it makes almost impossible to define a specific built-in strategy for Surface without the high risk of becoming obsolete in a couple of months. I believe the best approach would be to start with separate libs for that and if eventually, those libs require any change in Surface, we can define new extension points to make them work.

> What are your thoughts on this idea @msaraiva? I would be willing to work on a prototype to get some ideas flowing, with your guidance.

That's awesome! I have no idea right now which approach would be better for Surface so feel free to try any of them, following your judgment. You can try the "separate lib" strategy and let me know if you need any assistance. I'll be glad to help!

**Comment by Malvoz** [comment]
I saw the reference to this issue from the CSSWG thread on "bringing back scoped styles"; I just want to let you know there's a `scoped` attribute polyfill which might be of interest to you (created and maintained by a Googler): https://github.com/samthor/scoped

**Comment by msaraiva** [comment]
That looks Interesting. Thanks for the link, @Malvoz.

**Comment by fpalluel** [comment]
> I am very much so excited about this.
> 
> Another take on styling could be following the approach of what [Styled Components](https://styled-components.com/) is for React. We could have this as a new library or as part of Surface.
> 
> One of the main benefits of styled-components is being able to have dynamic CSS, which can take component props and update the CSS accordingly, and allows you to run Elixir code to calculate CSS. This also opens up the style system to allow helper libraries from the community: These may look similar to https://polished.js.org/, and https://github.com/styled-system/styled-system
> 
> StyledSurface (or whatever you would want to call it 🙂) could learn from the internals of StyledComponents on when and how to append to the DOM with updated styles. We would then, if I am not mistaken could leverage LiveView to broadcast any style updates to a style tag at the bottom of the HTML head tag. The auto-generated classes would work the same as with the original scoped styles idea above.
> 
> What are your thoughts on this idea @msaraiva? I would be willing to work on a prototype to get some ideas flowing, with your guidance.

Hello, indeed this is most interesting ! I've been using Styled Component and React for some time and the  first thought I had when Surface was created is "If we only we could style Surface components in the same way !" ;-)

Have you started anything about this  ?
Thanks :-)

**Comment by msaraiva** [comment]
Hi @fpalluel!

Unfortunately, I wasn't able to start working on this yet as I'm still working on the core features for the first version. If you have already thoughts about the direction we should take regarding the API, feel free to share them. As I said, I'm not sure yet which direction we should take due to the uncountable number of different ways to work with CSS.

**Comment by fpalluel** [comment]
Hello @msaraiva :-)

Given the following `styled-components`JS code : 

```js
export const MyButton = styled.button`
      background-color:  ${props => (props.is_loading ? props.theme.background_color.loading : props.theme.background_color.primary)};
      border: none;
      color: ${props => props.theme.color.primary};
      padding: 15px 32px;
      text-align: center;
      text-decoration: none;
      display: inline-block;
      font-size: 34px;
`;
```

```js
export const MySpecialButton = styled.MyButton`
      font-size: 16px;
`;
```
Note: `theme` gets injected from parent component via the React's Context API (same name, same purpose ;-) )

In Surface this could be (naively) :

```elixir
defmodule SurfaceTestWeb.MyButton do
  use Surface.Component

  @doc "The loading state"
  property is_loading, :boolean, required: true

  @doc "The main content"
  slot default, required: true

  context get theme, from: Main

  def style do
    css """
      background-color: {{ if @is_loading, do: @theme.background_color.loading, else: @theme.background_color.primary }} ;
      border: none;
      color: {{@theme.color.primary}};
      padding: 15px 32px;
      text-align: center;
      text-decoration: none;
      display: inline-block;
      font-size: 16px;
    """
  end

  def render(assigns) do
    ~H"""
      <button>
        <slot />
      </button>
    """
  end
end
```
and 

```elixir
defmodule SurfaceTestWeb.MySpecialButton do
  use Surface.Component
  alias SurfaceTestWeb.MyButton

  @doc "The loading state"
  property is_loading, :boolean, required: true

  @doc "The main content"
  slot default, required: true

  def style do
    css """
      font-size: 34px;
    """
  end

  def render(assigns) do
    ~H"""
      <MyButton is_loading={{ @is_loading }}>
        <slot />
      </MyButton>
    """
  end
end
```
'style' being a optional callback and `css` a preprocessing macro.

Under the hood, for instance for `MySpecialButton`, the css would be parsed, the properties would be merged with `MyButton` css properties,  variables would be interpolated.
The final css is a unique class, which gets injected in a `style` tag, within the `header`, as @garretttaco suggested.

The whole `styled-components` process is well described in [this article](https://medium.com/styled-components/how-styled-components-works-618a69970421 ).

Whichever css integration method is chosen for Surface, we will need a CSS parser :-(
Perhaps porting [stylis](https://github.com/thysultan/stylis.js) to elixir, using nimble_parsec ?

**Comment by msaraiva** [comment]
Hi @fpalluel!

Thanks a lot for your suggestion. I'll take a closer look at `styled-components`. There are still a few things that need to understand better, e.g. how does it work with existing classes, how to style nested nodes, etc.

**Comment by msaraiva** [comment]
Hi folks!

I've been doing some experiments regarding scoped styles in the last few days and I'd like to share some of the results and finally have a first official proposal. The solution is basically what has been proposed by @fpalluel (thanks again, BTW).

Let's start with the requirements.

## Requirements

What do we need/want:

1. Define CSS styles that are scoped by component, i.e. styles defined for one component should not leak to other components. 1. Support updating styles dynamically based on runtime assigns' changes.
1. Extend the API to allow defining styles in the same file of the component
1. Allow defining the styles in a separated collocated file.
1. Create a default processor that can compile the styles into CSS and split the styles into:
    - Static styles (applies to all instances of a component)
    - Dynamic styles (applies only to an instance).
1. Support custom processors by implementing the `StyleProcessor` behaviour.

## The proposal

We propose the following additions to in the API.

### The `style` macro

Defines a style for a component. Either by defining it inline or in a collocated file.

Example (inline):

![image](https://user-images.githubusercontent.com/457237/87169147-eecc9a00-c2a5-11ea-8c35-a1ae3f4c2dbb.png)

### Dynamic styles

This is probably the only significant change we have compared with the original proposal from @fpalluel. Instead of allowing arbitrary changes in the style using interpolation with `{{ }}`, we restrict the scope of changes by injecting `CSS variables` with the dynamic values.

Example:

![image](https://user-images.githubusercontent.com/457237/87170414-af9f4880-c2a7-11ea-9a2a-6286822aafc8.png)

Regarding the mapping syntax, we can make it more flexible allowing accepting explicit keys and functions as values. Something like:

```elixir
style vars: [^count, "--my-color": &calculated_value/1] do
  ...
end
```

And `calculated_value` could be:

```elixir
def calculated_color(assigns) do
  colors = ["red", "green", "black"]
  index = rem(assigns.count, length(colors))
  Enum.at(colors, index)
end
```

Why not allowing `{{ }}` directly? 

Because:

1. It's more complex.  We end up introducing a "new" language which is basically `CSS` or `SCSS` with `{{...}}` interpolation. The problem of introducing a new language to solve a problem is that the new language comes with its own set of problems, e.g. efficiently parsing the new rules related to interpolation, handling conflicting grammar between both languages (elixir and css/scss) and so on.
2. We have the concept of dynamic [CSS Variables](https://www.w3schools.com/css/css3_variables.asp) in `CSS`. We also have static variables in `SCSS` using the `$` prefix and interpolations using `#{}`. Why introduce another concept on top of all that? I believe mapping values from one language to the other is a safer/better approach.
3. Tooling support. By not creating a new language, we can reuse existing tools like editor support for auto-complete without breaking them. In the example above the syntax-highlighting works flawless, after all, it's valid `scss`.

### External files

By default, we could accept a collocated `.scss` file that would be automatically processed. Adding an explicit `style` would only be necessary if you need to pass any option, e.g. mapping dynamic values.

```elixir
style vars: ["--my-color": ^color]
```

We could also accept defining the file explicitly:

```elixir
style src: "./other_file.scss"
```
### Custom processors

As I said on other occasions, the CSS world moves extremely fast. My plan is to provide a default style processor for `scss`. However, other devs will be able to roll their own solution by implementing a `Surface.StyleProcessor`.

### Initial implementation

As an experiment, I've been working on an initial implementation for this proposal. 

How does it work?

Given the following `Counter` component:

```elixir
defmodule Counter do
  use Surface.LiveComponent

  data count, :integer, default: 32
  data color, :string, default: "black"

  style vars: ["--count": ^count, "--color": ^color] do
    scss """
    & {
      border: solid 2px var(--color);
      border-radius: 10px;
      padding: 10px;

      .title {
        font-size: calc(var(--count) * 1px);
      }

      > .subtitle {
        color: var(--color);
        font-weight: bold;
        font-style: italic;
        &:hover {
          color: orange;
        }
      }
    }
    """
  end

  def render(assigns) do
    ~H"""
    <div>
      <div>
        <button class="button is-info" :on-phx-click="dec">
          -
        </button>
        <button class="button is-info" :on-phx-click="inc">
          +
        </button>
      </div>
      <h1 class="title">font-size: {{ @count }}</h1>
      <h1 class="subtitle">color: {{ @color }}</h1>
    </div>
    """
  end

  def handle_event("inc", _, socket) do
    count = socket.assigns.count + 1
    {:noreply, assign(socket, count: count, color: get_color(count))}
  end

  def handle_event("dec", _, socket) do
    count = socket.assigns.count - 1
    {:noreply, assign(socket, count: count, color: get_color(count))}
  end

  def get_color(count) do
    colors = ["red", "green", "black"]
    index = rem(count, length(colors))
    Enum.at(colors, index)
  end
end
```

And adding a couple of instances to a page:

![image](https://user-images.githubusercontent.com/457237/87177088-723fb880-c2b1-11ea-897d-bf6a103b1ba7.png)

The actions taken will be:

1. The `scss` code is compiled by [LibSass](https://sass-lang.com/libsass) using [sass.ex](https://github.com/scottdavis/sass.ex).
1. The resulting `CSS` is split into two parts, one for the static `css` and the other for the dynamic part. 
1. All static parts from all components are merged into a `components.css` file currently located in `priv/static/css/`.
1. Each dynamic part is rendered along with its related component instance.

Here's the static file generated for the `Counter` component:

```css
/* This file was generated by the Surface compiler */

[data-sface-module="Counter"] {
  --count: 32;
  --color: black;
}

[data-sface-module="Counter"] {
  border: solid 2px var(--color);
  border-radius: 10px;
  padding: 10px;
}

[data-sface-module="Counter"] .title {
  font-size: calc(var(--count) * 1px);
}

[data-sface-module="Counter"] > .subtitle {
  color: var(--color);
  font-weight: bold;
  font-style: italic;
}

[data-sface-module="Counter"] > .subtitle:hover {
  color: orange;
}
``` 

Besides including all static properties into the `components.css`, we also initialize all dynamic properties with their default values, if any.

Here's the dynamic part:

![image](https://user-images.githubusercontent.com/457237/87179163-a9fc2f80-c2b4-11ea-987e-44e3461942fc.png)

### Limitations

* Although all modern browsers already have support for `CSS Variables`, there's no built-in support for IE.
* [sass.ex](https://github.com/scottdavis/sass.ex) hasn't been updated for a couple of years. We'll need to either try to update it or choose another lib.
* More? Tell me, please :)

In case you have any question or suggestion, please, let me know.

Cheers.

**Comment by aleDsz** [comment]
This is interesting, I was thinking about using something like structs for all possible values and parse it using those macros (`css`, `scss`, `sass`) and use a validator (maybe the same approach from Ecto) and returns an error in compile-time if some property has the wrong value.

This idea came from repo (https://github.com/davesnx/styled-ppx), which my friend constributed with a parser from CSS specs (https://github.com/EduardoRFS/reason-css-parser) for ReasonML and was implemented inside the StyledPPX library.

So I though about "Why not made it work with Surface, bringing a new experience when developing frontend with Phoenix?"

I'll let you know when I start to work with this project to improve :)

**Comment by msaraiva** [comment]
Hey @aleDsz!

Although the default built-in style processor will probably only handle `CSS` and `SCSS` code, I think the API should be flexible enough to allow something like:

```elixir
style :my_custom_style do
  # Any code allowed by the custom processor
  %MyStruct{
    ...
  }
end
```

The block content (along with any option defined) would be forwarded to the custom processor that is registered as `:my_custom_style`. This way the code can be validated and parsed by the processor at compile-time.

The `StyleProcessor` behaviour will be something like:

```elixir
```
In case you already have an idea of how the styles would look like using your elixir structs, you can send a couple of snippets so we can see if the proposed model will be a good fit.

**Comment by msaraiva** [comment]
In case anyone wants to give it a try, a `PoC` project can be found at https://github.com/msaraiva/surface_app/tree/style-experiments.

**Comment by aleDsz** [comment]
Hey @msaraiva!

My idea is not only to use structs to represent the types of properties, but also the values. However, instead of writing the style with structs, you would use the standard CSS parser for a typed CSS and do the validations in compile-time.

Like this example with ReasonML:

![](https://github.com/davesnx/styled-ppx/blob/master/docs/images/demo.png?raw=true)

You can write pure CSS, but the validation will occur when your are writing code, and if something it's wrong, we can make it shows an error in the editor.

![](https://github.com/davesnx/styled-ppx/blob/master/docs/images/compile-error.png?raw=true)

I was thinking re-use the idea from Ecto validations, to check if property support some value and create a message when the CSS have unsupported values.

Next week I'm on vacation from job, so I'll have more time to think about this idea and start code to see how it should work

**Comment by msaraiva** [comment]
Hi @aleDsz!

> You can write pure CSS, but the validation will occur when your are writing code, and if something it's wrong, we can make it shows an error in the editor.

Since all translation happens at compile-time, we already show any error found by the `Sass` compiler. Here's an example:

![image](https://user-images.githubusercontent.com/457237/87860295-a69b2080-c912-11ea-8000-c1b8d7700bf7.png)

> I was thinking re-use the idea from Ecto validations, to check if property support some value and create a message when the CSS have unsupported values.

Such validation can certainly be provided as soon as we have a parser or if `libsass` provides an API to access the parsed nodes. As far I as I know it doesn't, we just have access to the final generated `CSS`. Having our own parser would solve that.

> Next week I'm on vacation from job, so I'll have more time to think about this idea and start code to see how it should work

Awesome! Let me know if you have any question or need any assistance.

**Comment by aleDsz** [comment]
Yeah, sure.

Interesting this `SASS` validation, but as you said, if we create our own parser, we can't catch the error before generating the final `CSS`.

I have some ideas how to implement it, but first we need to create a parser from CSS specs, and after that, create the structs to implement the validator and such things.

**Comment by mxrguspxrt** [comment]
Good day! What is status with task and perhaps I can help anyhow with my limited knowledge, if task is well defined. :)

**Comment by msaraiva** [comment]
Hi @mxrguspxrt!

We're currently working on a few modifications in the context API in order to make it more flexible and more importantly, make it work nicely with LV's change tracking mechanism. We realized change tracking was being disabled whenever using contexts or slots. 

Since it involves changes in the API, we decided to address those issues before releasing our first stable version. The work is almost done and as soon as we release `v0.1.0`, we'll be back to the issue tracker and review the priorities. 

We might include this feature for `v0.2.0` if we feel we have enough information that can give us a clear picture of what should be done. I don't feel we're quite there yet. We didn't get much feedback regarding the proposed solution. I'd love to hear more about the pros and cons, as well as alternative solutions before adding this feature to the core lib.

Is the proposed solution using `LibSass` + `CSS variables` for dynamic styles a good fit for you?

**Comment by reywright** [comment]
I have minimal exposure to Phoenix and Elixir however I’m an experienced frontend dev I’d love to use this feature in an ongoing project I’m having developed. Some general thoughts:

- It would be nice if css can be autoprefixed out of the box as it’s a common thing done.
- I think SCSS support is generally more than sufficient as a MVP, though later it might be nice to provide easy configurable support for more bleeding edge spec stuff (similar to how babel does) without making the user entirely roll their own.
- Given that CSS is scoped, is there any plans to add a precompile step to generate a separate CSS file? If you have a site that renders 30 buttons or so, repeating the style block for each one, vs generating a stylesheet that can be cached across renders could be much more performing. 
- I have experience in production with styled components (otherwise known as css-in-is) and would really appreciate an implementation that doesn’t use it as it suffers from performance issues and is generally built for declarative frontend frameworks (and using things like polished and these other javascript frameworks in this framework don’t really make sense to me. the whole deal with phoenix live view is minimalism in how much javascript you need, so to take these javascript libraries for customizing saturation just seem a bit silly to me vs doing it in phoenix / elixir and passing the data from the controller or wherever onto these specific components.
- Assuming you have a single cached stylesheet, you may need to re-think how you do the dynamic values? It could be the case that css variables are scoped to some element but I’m not entirely sure that plays well with a global stylesheet? I don’t see the dynamic values being that big of an issue though (at least for version 1).

**Comment by msaraiva** [comment]
Hi @reywright!

Thanks for sharing your thoughts.

> Given that CSS is scoped, is there any plans to add a precompile step to generate a separate CSS file?

The PoC we have does exactly that. It generates a file containing all the styles scoped by each component/module. Only the dynamic parts are duplicated for each instance of a component.

**Comment by grncdr** [comment]
I'm not (yet) using Phoenix developer, just curious about this project, so this question might be very basic. How would a component like `<SomeItem selected={{ @foo }} />` (where the selected prop toggles the item color) be implemented?

**Comment by margusp2heathmont** [comment]
@msaraiva 

I hope you are good! :)

What are the future plans with https://github.com/msaraiva/surface/compare/scoped-styles  and when it might become part of master?

**Comment by Eein** [comment]
Another idea if you dont take a scss route that i was discussing with coworkers is potentially making styles composable for example

```elixir
defmodule Button do
  use Surface.Component

  def render(assign) do
    ~H"""
    <button>{{ @name }}</button>
    """
  end

end

defmodule Form do
  use Surface.Component

  @CancelButton styled Button, """
    color: red;
  """

  @SubmitButton styled Button, """
    color: blue;
  """

  def render(assigns) do
    ~H"""
    <div>
      <@CancelButton name="Cancel" />
      <@SubmitButton name="Cancel" />
    </div>
    """
  end
end
```

There may be a better way to do something like that - or i'm blind and glossed over it :)
Just some feedback! Great work so far!

**Comment by msaraiva** [comment]
Hi @Eein!

Thanks for the feedback. 

I like the idea. But what if we had a more generic way to create this kind of private lightweight stateless components (we can call them templates for now). The difference is that it would be able to set not only styles but any defined property. Something like:

```elixir
defmodule Form do
  use Surface.Component

  template CancelButton :: Button,
    label: "Cancel",
    style: """
    color: red;
    """

  def render(assigns) do
    ~H"""
    <div>
      ...
      <CancelButton/>
    </div>
    """
  end
end
```

We could even use this very same model to extend HTML tags directly. Like:

```elixir
defmodule Form do
  use Surface.Component

  template CancelButton :: button,
    class: "btn rounded",
    style: "color: red",
    do: ~H(Cancel)

  def render(assigns) do
    ~H"""
    <div>
      ...
      <CancelButton/>
    </div>
    """
  end
end
```

One great benefit of this approach is that, in both cases, we don't have to create real modules for them. They are just templates, which means that `<CancelButton/>` would be replaced by `<button class="btn rounded" ...>Cancel</button>` at compile-time.

WDYT?

**Comment by lewisf** [comment]
Curious ... given the original PoC relied on the now defunct translator to process styles, how would this work with the move to the compiler?

**Comment by msaraiva** [comment]
Hi @lewisf!

I'll move all code from the PoC to a new separate project and update it accordingly.

**Comment by Eein** [comment]
> WDYT?

looks great!
:+1:

**Comment by lewisf** [comment]
hi @msaraiva!

looking to see if i can help out with anything over the holidays / next couple of weeks. i'm REALLY excited to have style definitions co-located in Surface components -- there's something really nice about defining behavior and look and feel in the same place.

I was wondering if it's worth having the style definitions respond to changes in props/state. `styled-components` from the React ecosystem allows folks to do :

```javascript
const CancelButton = styled("button")`
  color: red,
  ${(props) => {
     // return a styles object
  }
`
```

With this ability to "extend" the css of any component, there isn't really a need for `class` anymore, which is kind of nice.

**Comment by Adzz** [comment]
Very excited about this thread! One question I can't tell so far, would the generated CSS classes be globally unique? I think that's a huge benefit of Styled Components for example.

**Comment by ElMassimo** [comment]
The proposed `style` approach, that can optionally map properties from the component to CSS variables, seems like a sound one.

A `styled-component`-like library can be easily built on top of that foundation, but it doesn't seem core to the project.

**Comment by msaraiva** [comment]
> looking to see if i can help out with anything over the holidays / next couple of weeks. i'm REALLY excited to have style definitions co-located in Surface components -- there's something really nice about defining behavior and look and feel in the same place.

Hi @lewisf! 

Sorry for the late response. As I mentioned before, anything regarding styles will probably be done in a separate project anyway so feel free to try out and bring for discussion any experiment you might have using whatever ideas you think are appropriate (although I believe your couple of weeks are gone 🙂). 

I still need to update the PoC we have to make it work with the latest changes. I'll do that after releasing `v0.1.2`.

> Very excited about this thread! One question I can't tell so far, would the generated CSS classes be globally unique? I think that's a huge benefit of Styled Components for example.

@Adzz yes. That's the idea.

**Comment by margusp2heathmont** [comment]
Scoped CSS style component approach basically addresses issue of hard to maintain CSS and conflicting CSS in big projects.

Same result, with different approach, could be achieved by using:

https://adamwathan.me/css-utility-classes-and-separation-of-concerns/
http://tailwindcss.com/

TailwindCSS configuration simply enables custom configuration, if needed, but defaults are also very good.

**Comment by Blond11516** [comment]
Hi @msaraiva!

I want to say I really like your suggestion as a first step to bring scoped CSS into Surface. Leveraging SASS and CSS variables is a great idea and will help get started much quicker. That said I see a few issues I think are worth mentioning.

_Disclaimer_: I'm a full-time React developer and a big fan of [styled-components](https://styled-components.com/). As a whole, my suggestions basically boils down to "please re-implement styled-components in Elixir", but it think they still have value when taken individually :slightly_smiling_face:

First, while CSS variables are great, they're not as versatile as a full-blown templating solution. Their main drawback is that they can't be used outside of property values. Most of the time this is fine, but it can be an issue, namely in media queries. For example, with React and styled-components, I could do the following:
```javascript
const Wrapper = styled.div`
  @media (max-width: ${props => props.mobileBreakpoint}) {
    ...
  }
`
```
There is no way to reproduce this with CSS variables. Again, this is still a great way to get started quickly, but I think in the long run full templating support will be a really nice feature to have, if not a must. Templating also has the advantage that if a variable needs to be calculated, you can keep the calculation inside the CSS (assuming it's fairly short), thus maintaining a better separation of concerns.

Second, I think having the ability to compose styles is very important (again, more of a long term goal), as @Eein suggested. I'm not sure if this is just an implementation detail or if you had a specific reason for it, but the `data-sface-module` makes that kinda hard I think. The way styled-components works is that it simply generates and injects classes into components. This means that composition is simply a matter of adding a new class on top of the already existing ones. Just something to consider.

Second-and-a-half, I'm not a big fan of having a single `style` macro that has to be applied to the whole module. I would personally much prefer to have the lib generate components for me that I can then use in my markup as I see fit. This has the advantage of making the markup/styling clearer because you don't have to add `class` all over the place. It also makes it easier to figure out which element uses which style because you don't have to look for the `class` attributes which can get a little lost if your markup uses many props. (I marked this as "second-and-a-half" because this point is kind of a prerequisite to the composition solution as I see it)

Third, regarding the complexity of the solution I'm proposing, losing compile time validation of CSS is really not a big deal to me. Unlike buisness logic, which has many corners and edge cases that compilation errors can greatly help to deal with, CSS is imo more "one-time-deal" type of thing. What I mean is that when you're writing CSS, you usually have the browser open on the side and are _constantly_ checking to see if what you've written does what you want. Actually, a great way to write CSS is to actually write it in the dev tools and only copy-paste it into your project once it's done. Besides, browser dev tools are already really good at telling if your CSS is valid or not.

Finally, I want to mention that the big drawback with what I'm proposing is that collocated stylesheets are kind of out of the question. This isn't a problem for me as I honestly hate the idea of having to navigate between three files to access my markup, logic and style, but I do realize that many people in the Phoenix community seem to really like it, so it's definitely something to consider.

**Comment by msaraiva** [comment]
Hi @Blond11516!

Thanks for your input. Those are very good points to consider and shows exactly how hard is to pick a specific strategy to implement as a built-in feature in Surface. There are too many ways to use CSS in a component-based project today, each one of them providing different tradeoffs. I'm still not convinced that we should bring all that mess to the core.

**Comment by Blond11516** [comment]
Agreed, an external library is a much better fit for this. Besides that leaves the door open for others to do their own thing, so I can have my styled-components clone while the rest of the community uses collocated styles! :D

**Comment by clsource** [comment]
Just my 2 cents.
Since this will not be in the core due to the different approaches possible.

Maybe something can be configured using the Javascript Hooks to apply styles 
maybe using https://github.com/indooorsman/esbuild-css-modules-plugin

or https://cssinjs.org/?v=v10.9.0

**Comment by msaraiva** [comment]
We have a new implementation for this feature based on Vue's approach. See https://github.com/surface-ui/surface/pull/621 for details.

**Comment by msaraiva** [comment]
Implemented in #621.
